### PR TITLE
fix(complexity): stop using dotted names as a proxy for signature-only stubs

### LINKED ISSUE #1922: WASM complexity engine skips entire file when every function has a dotted name (Lua module pattern, Go receiver-only files)
## Summary

`hasFuncBody()` in `src/ast-analysis/engine.ts` excludes any definition whose `name` contains a dot (`!d.name.includes('.')`), intended to filter out bodyless TS/interface method signatures extracted with dotted names (e.g. `InterfaceName.methodName`). This heuristic is also applied as the **file-level gate** deciding whether the WASM complexity visitor runs at all:

```js
function hasFuncBody(d) {
  return (
    (d.kind === 'function' || d.kind === 'method') &&
    d.line > 0 &&
    d.endLine != null &&
    d.endLine > d.line &&
    !d.name.includes('.')
  );
}
```

```js
const needsWasmComplexity = defs.some((d) => hasFuncBody(d) && !d.complexity);
```

For **module-table languages** (most visibly Lua's idiomatic `local M = {}; function M.foo() ... end; return M` pattern), every top-level function in a file can legitimately have a dotted name and a real multi-line body — these are not signature stubs. When *every* definition in a file is dotted, `defs.some(...)` is false for the whole file, so the WASM complexity visitor never runs and **no function in that file gets complexity data**, even though each one has a real, non-trivial body.

Once the visitor *does* run (triggered by at least one non-dotted def in the same file), `storeComplexityResults` maps results back to definitions purely by line number — it does not re-apply the dotted-name filter — so individual dotted-name functions are computed correctly. The bug is specifically the file-level "does this file need complexity at all" gate.

## Evidence

Discovered while fixing #1782 (Lua missing complexity rules entirely). After adding Lua to `COMPLEXITY_RULES`/`HALSTEAD_RULES`, comparing native vs WASM-built graphs for the same fixtures:

```
$ codegraph complexity --file tests/benchmarks/resolution/fixtures/lua/repository.lua --health -T --json --engine native
functions: 4   (M.delete, M.count, M.save, M.find_by_id)

$ codegraph complexity --file tests/benchmarks/resolution/fixtures/lua/repository.lua --health -T --json --engine wasm
functions: 0
```

Same for `tests/benchmarks/resolution/fixtures/lua/service.lua` (M.create_user, M.get_user, M.remove_user, M.summary) — 4 via native, 0 via WASM.

Per repo convention (CLAUDE.md): "Both engines must produce identical results. If they diverge, the less-accurate engine has a bug." Native computes complexity unconditionally at extraction time (bypassing this JS-side gate entirely), so it's unaffected — WASM is the less-accurate engine here.

## Cross-language risk

The same dotted-naming pattern exists for Go receiver methods (`func (r *Repo) Save() {}` → extracted as `"Repo.Save"`, kind `method`, real body) — confirmed via `src/extractors/go.ts:86`. Go fixtures didn't reproduce the zero-functions symptom only because the affected test files also contain at least one non-dotted free function (e.g. `NewUserService`), which is enough to flip `defs.some(...)` true for the whole file and let the visitor run. A Go file containing *only* receiver methods (no free functions) would hit the same zero-functions bug.

## Suggested fix direction

The dotted-name check conflates two different things:
1. "Is this a real function with a body?" — already covered by `d.endLine > d.line` (a multi-line span).
2. "Is this a TS/interface method signature that happens to span multiple lines due to formatting?" — the actual scenario the dot check was added for (see the `#846` reference near `initWasmParsersIfNeeded`).

A fix likely needs to stop using name-shape as a proxy for "no body" — e.g. relying on a more direct signal from the extractor (a `hasBody`/`isSignature` flag on the definition) rather than inferring it from the presence of a dot in the name, since dots are also a legitimate namespacing convention (Lua module tables, Go receivers) unrelated to whether a body exists. This needs verification across all extractors that produce dotted names (TS interfaces/ambient declarations, Go, Lua, possibly others) to avoid regressing the original interface-stub-filtering behavior.

## Scope note

Not fixed as part of #1782 — that issue is scoped to Lua's *missing* complexity-rules configuration (both engines had zero Lua support at all). This is a separate, pre-existing heuristic bug in the shared WASM gating logic, unmasked only after fixing #1782 (previously Lua's complexity was 100% broken on both engines, so this narrower gap wasn't independently observable). Filing separately per repo scope-discipline convention rather than expanding that PR.

### PR DESCRIPTION
**Pull request by carlos-alm**
## Summary

`hasFuncBody()` in `src/ast-analysis/apply-results.ts` (plus duplicated copies of the same check in `features/complexity.ts` and `features/cfg.ts`) excluded any definition whose `name` contained a dot, intended to filter out bodyless TS interface method signatures (`InterfaceName.methodName`).

Dotted names are also the **normal qualified name** for every real, bodied method across the entire codebase's extractors: Lua's `M.foo` module-table pattern, Go/Java/C#/PHP/Rust receiver or impl methods, and plain `Class.method` in any OOP language. The dot check was a proxy that happened to be true for the one case that motivated it, but false for everything else that uses `.` as a namespacing convention.

Since the file-level "does this file need WASM complexity" gate is `defs.some(hasFuncBody)`, a file where **every** function/method definition happens to have a dotted name (a Lua module file, a Go file with only receiver methods, a Python/Java/etc. file with a single class and no free functions) made the gate false for the *entire file* — none of its functions got complexity or CFG data via WASM.

## Fix

Replaced the name-shape heuristic with `Definition.bodyless`, a direct signal each extractor sets from the AST node's actual body field, at the specific sites verified (via each language's `node-types.json`) to produce genuinely bodyless dotted definitions:

- TS/JS: `method_signature` (interface stub) — `extractInterfaceMethods`
- Go: `method_elem` (interface stub, no `body` field at all) — `handleGoInterfaceType`; also defensively marks external/asm-linked receiver methods (`method_declaration` with an optional `body` field) in `handleGoMethodDecl`
- Java: interface methods (`extractJavaInterfaceMethods`) and abstract class methods (`handleJavaMethodDecl`, both use `method_declaration`, body optional)
- C#: interface methods (`handleCsInterfaceDecl`) and abstract class methods (`handleCsMethodDecl`)
- PHP: interface methods (`handlePhpInterfaceDecl`) and abstract class methods (`handlePhpMethodDecl`)
- Rust: trait `function_signature_item` (required, no body field) vs. `function_item` (default impl, body required) — `extractTraitMethods`

Python, Ruby, and Lua don't need this: their grammars structurally require a body on every function/method definition (even `pass`/`...`/`end`), so there's no bodyless-but-dotted case to guard against.

`hasFuncBody`, `initWasmParsersIfNeeded`, and `classifyDefinitionForNativeBulk` (`features/complexity.ts`) and `hasNativeCfgForFile` (`features/cfg.ts`) now check `!d.bodyless`. The latter two previously duplicated the dot-check logic inline — folded onto the shared `hasFuncBody` instead, removing the duplication.

**Native (Rust) mirror:** native already computed complexity unconditionally at extraction time (ignoring dots entirely), so the original bug never affected it directly. But `allNativeDataComplete`/`detectNativeNeeds` in `ast-analysis/engine.ts` evaluate `hasFuncBody` against native-bridged `Definition` objects too (from `parseFilesFull`) — without a native-side `bodyless` signal, those interface/trait/abstract stub definitions would look identical to WASM's pre-fix "possibly-needs-complexity" defs to the JS-side gate, silently forcing an unnecessary WASM/JS fallback re-parse on every native build touching an interface, trait, or abstract method (the exact perf regression #606 originally fixed, reintroduced through the native path). Added the same `bodyless` field + extraction-site logic to the native `Definition` struct and every extractor listed above.

## Testing

- `tests/unit/apply-results.test.ts` — updated `hasFuncBody` unit tests (dotted-with-body is now `true`; explicit `bodyless: true` is `false`)
- `tests/integration/issue-1922-wasm-complexity-dotted-names.test.ts` — new integration test building a Lua module-table file, a Go receiver-only file, and a mixed TS file (real dotted class method + genuinely bodyless multi-line interface signature) via the WASM engine, asserting complexity is correctly computed/excluded
- 4 new/extended Rust unit tests (`go.rs`, `javascript.rs`, `rust_lang.rs`, `java.rs`) asserting `bodyless` on representative interface/trait/abstract/receiver-method fixtures
- Live-verified native vs. WASM parity (identical `functions`/`cyclomatic` output) for Lua, Go, Python, TypeScript, and Java fixtures reproducing the reported bug
- `npm test`: 255 files, 4027 passed (0 failed — one `tests/integration/check.test.ts` flake unrelated to this change and already tracked as #1990)
- `npm run lint`: clean (pre-existing unrelated warning in `tests/fixtures/issue-1833-plain-type-import/consumer.ts`)
- `cargo test` (codegraph-core): 607 passed, 0 failed
- `npm run build` / `cargo check`: clean

## Out-of-scope findings filed separately

While auditing every native extractor site that mirrors a WASM interface/trait stub extractor (to verify `bodyless` correctness and avoid regressing the original interface-stub-filtering behavior), found and filed three independent, pre-existing bugs — not fixed here to keep this PR scoped to the file-level complexity gate:

- #2053 — Native Rust computes trivial complexity for bodyless Rust trait methods (`function_signature_item`); WASM correctly omits them
- #2054 — Native PHP extractor double-extracts interface methods as a spurious bare-named definition (missing interface guard in `handle_method_decl`)
- #2055 — C# WASM complexity visitor computes trivial complexity for bodyless interface method signatures; native correctly omits them

Stacked on #1920 (base branch `fix/issue-1920-wasm-dynamic-import-destructure-extraction`).

Fixes #1922

### CODE REVIEW

File: tests/unit/apply-results.test.ts [hunk 1]
```diff
@@ -52,8 +52,25 @@ describe('hasFuncBody', () => {
...
+
+  it('is false when the extractor marks the definition bodyless (interface/trait/abstract signature)', () => {
+    expect(
+      hasFuncBody({ name: 'Foo.bar', kind: 'method', line: 5, endLine: 10, bodyless: true }),
+    ).toBe(false);
+    // Even a non-dotted signature-only declaration is excluded via `bodyless`.
+    expect(
+      hasFuncBody({ name: 'bar', kind: 'function', line: 5, endLine: 10, bodyless: true }),
+    ).toBe(false);
   });
```
[the remark this answers is not in the document]
Comment (carlos-alm) [reply tests/unit/apply-results.test.ts:72]: Fixed. Removed `name` from every `hasFuncBody({...})` object literal in this file — both the pre-existing calls at lines 42-52 (also flagged separately in the "Comments Outside Diff" section of the summary) and the newly-added ones in these two `it` blocks.

One deviation from the literal suggested diff: after stripping `name`, the two assertions in "is true for a dotted name with a real body..." became byte-identical (`hasFuncBody({ kind: 'method', line: 5, endLine: 10 })` twice) — the whole point of the original `Foo.bar` vs `M.foo` distinction was lost once `name` left the input entirely, since `hasFuncBody` no longer reads it at all. Kept a single assertion instead of two duplicate lines and adjusted the comment to explain why.

Verified `tests` is excluded from `tsconfig.json`'s `include` (`"include": ["src/**/*"]`, `"exclude": [...,"tests",...]`), so this is not currently breaking any CI check — confirmed with `npm run typecheck` (`tsc --noEmit`) passing before and after. Also directly verified the fix holds under a hypothetical stricter config: temporarily pointed `tsconfig.json` at this test file (include + drop the `tests` exclude) and ran `tsc --noEmit` — zero errors, then reverted the config.

**Comment by carlos-alm** [comment]
Addressed both findings from this review:

1. The inline comment on the two new `it` blocks (lines 55-74) — fixed, replied there directly.
2. The "Comments Outside Diff" finding on the pre-existing calls at lines 41-53 — fixed in the same pass, `name` removed from every `hasFuncBody({...})` literal in the file (9 call sites total).

Separately: this PR was stacked behind #2021-#2052, all of which have since been merged to `main`. I merged `origin/main` into this branch to resolve the resulting conflict, reconstructing this PR's own `bodyless` changes on top of main's current state file-by-file (verified each of the 47 files this commit touches against the reference diff to confirm nothing was dropped or duplicated by the merge). No functional changes beyond the merge and the two test-typing fixes above.

**Comment by carlos-alm** [comment]
@greptileai
